tests/core/pyspec/eth2spec/test/validator/test_validator_unittest.py

```diff
@@ -0,0 +1,296 @@
+        spec,
+        spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE * 2),
+        period_start,
+        success=True,
+    )
+    run_is_candidate_block(
+        spec,
+        spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE * 2 - 1),
+        period_start,
+        success=False,
+    )
+
+
+@with_all_phases
+@spec_state_test
+def test_get_eth1_data_default_vote(spec, state):
+    min_new_period_epochs = get_min_new_period_epochs(spec)
+    for _ in range(min_new_period_epochs):
+        next_epoch(spec, state)
+
+    state.eth1_data_votes = ()
+    eth1_chain = []
+    eth1_data = spec.get_eth1_vote(state, eth1_chain)
+    assert eth1_data == state.eth1_data
+
+
+@with_all_phases
+@spec_state_test
+def test_get_eth1_data_consensus_vote(spec, state):
+    min_new_period_epochs = get_min_new_period_epochs(spec)
+    for _ in range(min_new_period_epochs):
+        next_epoch(spec, state)
+
+    period_start = spec.voting_period_start_time(state)
+    votes_length = spec.get_current_epoch(state) % spec.EPOCHS_PER_ETH1_VOTING_PERIOD
+    state.eth1_data_votes = ()
+    eth1_chain = []
+    eth1_data_votes = []
+    block = spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE)
+    for i in range(votes_length):
+        eth1_chain.append(block)
+        eth1_data_votes.append(spec.get_eth1_data(block))
+
+    state.eth1_data_votes = eth1_data_votes
+    eth1_data = spec.get_eth1_vote(state, eth1_chain)
+    assert eth1_data.block_hash == block.hash_tree_root()
+
+
+@with_all_phases
+@spec_state_test
+def test_compute_new_state_root(spec, state):
+    pre_state = state.copy()
+    post_state = state.copy()
+    block = build_empty_block(spec, state, state.slot + 1)
+    state_root = spec.compute_new_state_root(state, block)
+
+    assert state_root != pre_state.hash_tree_root()
+
+    # dumb verification
+    spec.process_slots(post_state, block.slot)
+    spec.process_block(post_state, block)
+    assert state_root == post_state.hash_tree_root()
+
+
+@with_all_phases
+@spec_state_test
+def test_get_block_signature(spec, state):
+    privkey = privkeys[0]
+    pubkey = pubkeys[0]
+    block = build_empty_block(spec, state)
+    signature = spec.get_block_signature(state, block, privkey)
+    domain = spec.get_domain(state, spec.DOMAIN_BEACON_PROPOSER, spec.compute_epoch_at_slot(block.slot))
+    signing_root = spec.compute_signing_root(block, domain)
+    assert bls.Verify(pubkey, signing_root, signature)
+
+
+# Attesting
+
+
+@with_all_phases
+@spec_state_test
+def test_get_attestation_signature(spec, state):
+    privkey = privkeys[0]
+    pubkey = pubkeys[0]
+    attestation_data = spec.AttestationData(slot=10)
+    signature = spec.get_attestation_signature(state, attestation_data, privkey)
+    domain = spec.get_domain(state, spec.DOMAIN_BEACON_ATTESTER, attestation_data.target.epoch)
```

Comment:
could consider having all of these `test_get_*_signature` tests call a common helper `run_get_signature_test(state, obj, domain, get_signature_fn)` to de-duplicate similar code paths